torch/_dynamo/variables/dicts.py

```diff
@@ -1,45 +1,167 @@
 import collections
 import dataclasses
+import enum
 import functools
 import inspect
 import sys
 from typing import Any, Dict, List, Optional
 
-import torch
-import torch.fx
+from torch._subclasses.fake_tensor import is_fake
 
 from .. import variables
 from ..bytecode_transformation import create_call_function, create_instruction
 from ..eval_frame import skip_code
 
 from ..exc import unimplemented
 from ..guards import GuardBuilder, make_dupe_guard
-from ..source import AttrSource, GetItemSource, GlobalWeakRefSource
-from ..utils import global_key_name, istensor, iter_contains
+from ..source import AttrSource, GetItemSource
+from ..utils import iter_contains, specialize_symnode
 from .base import MutableLocal, VariableTracker
 from .constant import ConstantVariable
-from .tensor import TensorVariable
+
+
+# Note: [Adding a new supported class the keys of ConstDictVarialble]
+# You'll need to add it to:
+# - `is_hashable_python_var` in this file
+# - `is_hashable` in this file
+# - `const_repr` in util.py, and perhaps modify DICT_KEYS in guards.py
+
+
+def is_hashable_python_var(x):
+    from torch import Tensor
+
+    # Note: Keep me in sync with is_hashable!
+    # Even better, we should have a map of functions connecting the two
+    from ..allowed_functions import is_builtin_callable
+
+    return (
+        ConstantVariable.is_literal(x)
+        or isinstance(x, (Tensor, enum.Enum))
+        or is_builtin_callable(x)
+        or (isinstance(x, tuple) and all(is_hashable_python_var(e) for e in x))
+    )
+
+
+def is_hashable(x):
+    # Keep me in sync with is_hashable_python_var!
+    # Even better, we should have a map of functions connecting the two
+    if isinstance(x, variables.TensorVariable):
+        # Tensors are hashable if they have an example_value (a fake tensor)
+        # Most VT's should have one.
+        # It'd be nice if at some point we could assert that they all have one
+        return x.as_proxy().node.meta.get("example_value") is not None
+    elif isinstance(x, variables.TupleVariable):
+        return all(is_hashable(e) for e in x.items)
+    else:
+        return isinstance(
+            x,
+            (
+                variables.BuiltinVariable,
+                variables.SymNodeVariable,
+                variables.ConstantVariable,
+                variables.EnumVariable,
+            ),
+        )
 
 
 class ConstDictVariable(VariableTracker):
-    def __init__(self, items, user_cls, **kwargs):
+    class _HashableTracker:
+        """
+        Auxiliary opaque internal class that wraps a VariableTracker and makes it hashable
+        This should not be seen or touched by anything outside of ConstDictVariable and its children
+        Note that it's also fine to put VTs into dictionaries and sets, but doing so does not take into account aliasing
+        """
+
+        def __init__(self, vt):
+            # We specialize SymNodes
+            vt = specialize_symnode(vt)
```

Comment:
We specialize SymNodes if they are keys to a dictionary. We could do this a bit more lazily, but it would complicate the code a fair amount for not much value IMO.

The second point here is that `_HashableTracker.__init__` has side effects. For example, we also specialize a symbolic int `x` if we do `x in my_set`.